src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java

```diff
@@ -916,10 +933,61 @@ private static void applyRestrictionsToNewEdge(
                 "Recreate new restriction {} with split edge as to edge {}", splitTurnRestriction,
                 newEdge
         );
-        graph.addTurnRestriction(restriction.from, splitTurnRestriction);
+        restriction.from.addTurnRestriction(splitTurnRestriction);
         // Former turn restriction needs to be removed. Especially no only_turn
-        // restriction to a non existent edge must not survive
-        graph.removeTurnRestriction(restriction.from, restriction);
+        // restriction to a non-existent edge must not survive
+        restriction.from.removeTurnRestriction(restriction);
+    }
+
+    /**
+     * Add a {@link TurnRestriction} to this edge.
+     *
+     * This method is thread-safe as modifying the underlying set is synchronized.
+     */
+    public void addTurnRestriction(TurnRestriction turnRestriction) {
+        if (turnRestriction == null) { return; }
+        synchronized (this) {
```

Comment:
Synchronizing on `this` is NOT considered best-practice, but in this case I think it is our best option. The reason is that you are not synchronizing on the "thing" you are protecting, but blocking access to its owner. The correct way is to use a AtomicReference, but then we would use even more memory.

Reply:
My first version synchronized on `turnRestriction` but that is not final. I agree that `this` isn't great but it's the best alternative.